### PR TITLE
nodejs-express: support apps using socket.io and pino

### PR DESCRIPTION
**Pull request by sam-github**
Bumps stack to 0.3, because the app.js interface is different (it allows dependency injection).

Questions:
- should scaffold be modified to allow dep injection into its routes? ATM, if (for example) you had opened a db connection, you could (EDIT:) *not* pass that to all the route handlers, except maybe by assigning it into a global variable.
- is 0.3.0 the right version number?
- is it worth putting the chat-server as its own example app structure? I'm not sure how used they are. And if someone wanted socket.io AND the scaffold app, what then? Perhaps just in the README? Or maybe its obvious enough, no docs needed other than the comment I put in the template app.js?

### Checklist:

- [x] Read the [Code of Conduct](https://github.com/appsody/website/blob/master/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/appsody/website/blob/master/CONTRIBUTING.md).

- [x] Followed the [commit message guidelines](https://github.com/appsody/website/blob/master/CONTRIBUTING.md#commit-message-guidelines).

- [x] Stack adheres to [Appsody stack structure](https://github.com/appsody/website/blob/master/content/docs/stacks/stack-structure.md).

### Modifying an existing stack:

- [x] Updated the stack version in `stack.yaml`

commit b317ed5e9e5b524008f9bc7ddbc5c2e382f1c14f
Author: Sam Roberts <vieuxtech@gmail.com>
Date:   Thu Nov 21 13:53:30 2019 -0800

    WIP: use in-dev prometheus and dash until PRs are closed and released

    Blocked on:
    - https://github.com/CloudNativeJS/appmetrics-prometheus/pull/35
    - https://github.com/RuntimeTools/appmetrics-dash/pull/184

commit a3303bd4ed8938954a71c57d8d0080f18ec330cf
Author: Sam Roberts <vieuxtech@gmail.com>
Date:   Fri Nov 15 16:48:02 2019 -0800

    nodejs-express: expose server to the application

    This allows application specific configuration, such as adding
    socket.io, or logger middleware.

    socket.io example app.js:

            module.exports = (options) => {
              const io = require('socket.io')(options.server, {
                path: '/ws',
              });
              const app = require('express')()

              app.get('/', (req, res) => {
                res.send("Hello from Appsody!");
              });

              return app;
            };

commit 4b4aa8aefe8e41e32089926590c5c1bcb085adea
Author: Sam Roberts <vieuxtech@gmail.com>
Date:   Mon Nov 18 15:09:25 2019 -0800

    nodejs-express: use explicit APIs for dash and metrics

    Using .attach() from appmetrics-dash or appmetrics-prometheus monkey
    patches the node core http and https modules in a way that prevents
    appliations from using socket.io.

    Its unnecessary, use explicit APIs to enable these features.

commit 2c791267c120f627c4f3988664d0b24dd5c6c3c7
Author: Sam Roberts <vieuxtech@gmail.com>
Date:   Fri Nov 22 13:49:20 2019 -0800

    nodejs-express: end every line with a NL

    Git and github and many other tools complain when the last line of a
    file lacks the expected line ending.

Test with https://github.com/socketio/chat-example, which when rewritten to conform to appsody looks like:

app.js:
```
module.exports = (options) => {                                                  
  const io = require('socket.io')(options.server);                               
  const app = require('express')();                                              
                                                                                 
  app.get('/', function(req, res){                                               
    res.sendFile(__dirname + '/index.html');                                     
  });                                                                            
                                                                                 
  io.on('connection', function(socket){                                          
    socket.on('chat message', function(msg){                                     
      io.emit('chat message', msg);                                              
    });                                                                          
  });                                                                            
  return app;                                                                    
};                  
```
with index.html:
```
<!doctype html>                                                                  
<html>                                                                           
  <head>                                                                         
    <title>Socket.IO chat</title>                                                
    <style>                                                                      
      * { margin: 0; padding: 0; box-sizing: border-box; }                       
      body { font: 13px Helvetica, Arial; }                                      
      form { background: #000; padding: 3px; position: fixed; bottom: 0; width: 100%; }
      form input { border: 0; padding: 10px; width: 90%; margin-right: .5%; }    
      form button { width: 9%; background: rgb(130, 224, 255); border: none; padding: 10px; }
      #messages { list-style-type: none; margin: 0; padding: 0; }                
      #messages li { padding: 5px 10px; }                                        
      #messages li:nth-child(odd) { background: #eee; }                          
      #messages { margin-bottom: 40px }                                          
    </style>                                                                     
  </head>                                                                        
  <body>                                                                         
    <ul id="messages"></ul>                                                      
    <form action="">                                                             
      <input id="m" autocomplete="off" /><button>Send</button>                   
    </form>                                                                      
    <script src="/socket.io/socket.io.js"></script>                              
    <script src="https://code.jquery.com/jquery-1.11.1.js"></script>             
    <script>                                                                     
      $(function () {                                                            
        var socket = io();                                                       
        $('form').submit(function(){                                             
          socket.emit('chat message', $('#m').val());                            
          $('#m').val('');                                                       
          return false;                                                          
        });                                                                      
        socket.on('chat message', function(msg){                                 
          $('#messages').append($('<li>').text(msg));                            
          window.scrollTo(0, document.body.scrollHeight);                        
        });                                                                      
      });                                                                        
    </script>                                                                    
  </body>                                                                        
</html>                                                                         
```

### Related Issues:

Fixes https://github.com/appsody/stacks/issues/447

### CODE REVIEW

File: incubator/nodejs-express/image/project/package.json [hunk 1]
```diff
@@ -15,13 +15,13 @@
   },
   "dependencies": {
     "@cloudnative/health-connect": "^2.0.0",
-    "appmetrics-prometheus": "^3.0.0",
+    "appmetrics-prometheus": "github:sam-github/appmetrics-prometheus#optional-patching",
```
Comment (Kamran64) [inline incubator/nodejs-express/image/project/package.json]: Hey @sam-github, spoke with @andrewhughes101 about this PR. Let's wait until https://github.com/CloudNativeJS/appmetrics-prometheus/pull/35 is in from the appmetrics prometheus side so we can use an officially released version here.

File: incubator/nodejs-express/image/project/package.json [hunk 2]
```diff
@@ -15,13 +15,13 @@
...
     "express": "~4.17.1"
   },
   "devDependencies": {
+    "appmetrics-dash": "github:Runtimetools/appmetrics-dash#avoid-ws-conflict",
```
Comment (Kamran64) [inline incubator/nodejs-express/image/project/package.json]: Same here as above.

**Comment by sam-github** [comment]
This is a ***WIP*** - please do *not* merge it:

> WIP: use in-dev prometheus and dash until PRs are closed and released
> Blocked on:
> - https://github.com/CloudNativeJS/appmetrics-prometheus/pull/35
> - https://github.com/RuntimeTools/appmetrics-dash/pull/184

@karmran64 Any comment on the open questions?

- should scaffold be modified to allow dep injection into its routes? ATM, if (for example) you had opened a db connection, you could (EDIT:) not pass that to all the route handlers, except maybe by assigning it into a global variable.
- is 0.3.0 the right version number?
- is it worth putting the chat-server as its own example app structure? I'm not sure how used they are. And if someone wanted socket.io AND the scaffold app, what then? Perhaps just in the README? Or maybe its obvious enough, no docs needed other than the comment I put in the template app.js?

**Comment by Kamran64** [comment]
@sam-github - So I think most of these decisions should be down to the stack maintainers but to me, bumping to 0.3.0 sounds like the right call. I agree, the chat server should only be put in an example app if it's used enough. However it would be nice if the README's were expanded to detail some of the new additions.

**Comment by sam-github** [comment]
OK, extended README to describe the contract between the stack and the app.

Still waiting for upstream to republish the deps.

**Comment by sam-github** [comment]
Pushed a new commit:

        nodejs-express: add pino logger and middleware

Still waiting on upstream:
- https://www.npmjs.com/package/appmetrics-prometheus needs republish, @tobespc and @sjanuary have publish rights
- https://github.com/RuntimeTools/appmetrics-dash/pull/184 needs merge and republish, working on with @mattcolgate

**Comment by sam-github** [comment]
I'm still running `appsody stack validate` locally, but I think this is updated and ready to go -- the deps have been published, package-lock updated, etc.

@andrewhughes101 do you want to take a last look?

**Comment by andrewhughes101** [comment]
@sam-github Everything looks good to me

**Comment by sam-github** [comment]
@Kamran64 pls review, upstream has been published, should be ready to go

**Pull request merged**